### DISCUSSION TITLE
I Want to have an event handler for the browser's back button with next.js 13.5 (App router)

### DISCUSSION THREAD

**Discussion by sohitkumar** [vercel/next.js#63577] [Help]
### Summary

What I exactly want is if the user clicks on web browser back button, then redirect to homepage...

Suppose I am in Payment page, now if the user clicks on web browser's back button then redirect to homepage.

### Additional information

```js
I got one solution earlier, but it was very old and it's not supported by the router imported from next/navigation...

old code which is not working
useEffect(() => {
    router.beforePopState(({ as }) => {
        if (as !== router.asPath) {
            // Will run when leaving the current page; on back/forward actions
            // Add your logic here, like toggling the modal state
        }
        return true;
    });

    return () => {
        router.beforePopState(() => true);
    };
}, [router]);
```

**Comment by SouthLink** [comment]
Our application also encountered this problem, although it seems simple, it is actually a little difficult, because of the limitations of chrome, as well as the internal limitations of next.js。

To insert a history record of a specified link in the next.js framework and to support the user clicking the browser back button to redirect to the specified page, three requirements need to be met:
1. you need to use the history API to insert a browsing history.
2. the inserted history object must meet the internal requirements of next.js (next.js internally determines whether the history is "valid" or not by using the __N field)
[__N](https://github.com/vercel/next.js/blob/6b6590592eb95990cae768c099002d2b45359dbb/packages/next/src/shared/lib/router/router.ts#L929)
3. the user needs to interact with the page to activate the history entry. [when_popstate_is_sent](https://developer.mozilla.org/en-US/docs/Web/API/Window/popstate_event#when_popstate_is_sent)

We can do this:

```
const createNextHistory = (url, as, options = {}) => {
  return {
    url,
    as: as || url,
    options: { ...options },
    __N: true,
    key: Math.random().toString(36).slice(2, 10),
  };
};

const insertHistory = () => {
  let pathname = location.pathname;
  let search = location.search;
  let _href = location.href;

  // First insert a history of the home page
  history.replaceState(createNextHistory("/"), "", location.origin + "/m/");
  
 // Replace the url of the current page back to its original state
  history.pushState(createNextHistory(`${pathname}${search}`), "", _href);
};

```

This way, after the user has had a valid interaction on the page, clicking the browser's back button will redirect you to the page you want (this is true in chrome, it should work straight away in safari)

Note that this doesn't always work and may not work with next.js updates, I'm currently using version 13.2.0！

**Comment by icyJoseph** [comment]
Shouldn't you guys be able to use the: https://developer.mozilla.org/en-US/docs/Web/API/Window/popstate_event event?

**Reply by Stan-Stani** [reply]
Somehow my component unmounts before that event even gets triggered, which indicates that somehow Next.js knows I clicked the back button before the browser sends that event? Does it shim it or something?

**Reply by Stan-Stani** [reply]
Ah, apparently that event happens really late. https://developer.mozilla.org/en-US/docs/Web/API/Window/popstate_event
